[DS-3906] REST endpoint for discovering withdrawn and private items.

mspalti opened the pull request:
This small PR gives administrators the ability to discover (both) withdrawn and private items. It includes:

- A new discovery configuration (in discovery.xml) called "showUnDiscoverableItems". The configuration includes solr filters that find items that have been withdrawn or made private.
- A new `SolrServiceSearchPlugin` that prevents discovery of withdrawn or private items by non-administrative users.
- Changes to `DiscoveryRestRepository` that cause the `SolrServiceImpl.search()` method to be called with `includeUnDiscoverable = true` for requests that include the "showUnDiscoverableItems" configuration.
- Modification of `DiscoveryConfigurationService` to allow for discovery configurations that include the a Community handle (e.g.: "showUnDiscoverableItems.12345/12")

Integration tests will be added.

### Needed Feedback
It would be great to have feedback on how well the requirements described in the jira issue https://jira.duraspace.org/browse/DS-3906 have been met by this initial PR. If the general approach looks good I can refine things a bit and add ITs.  Some documentation should also be added to the REST contract, I think.

Also, browse for withdrawn items and private items are separate features on the development spreadsheet, but there's a note asking whether these features can be combined.  This configuration as of now combines the two.

mspalti committed:
[DS-3906] REST endpoint for discovering withdrawn and private items.

mspalti committed:
Fixing tests for travis (currently unable to test on local machine)

mspalti committed:
Formatting corrections.

benbosman commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/DiscoveryRestRepository.java:
```diff
@@ -102,7 +102,13 @@ public SearchResultsRest getSearchObjects(final String query, final String dsoTy
                 .buildQuery(context, scopeObject, discoveryConfiguration, query, searchFilters, dsoType, page);
-            searchResult = searchService.search(context, scopeObject, discoverQuery);
+
+            if (configuration != null && configuration.contentEquals("showUnDiscoverableItems")) {
```
I'd prefer to minimize the amount of "exceptions" where a different method is used based on the name of the configuration because it creates code which is hard to maintain or debug.
It would be much cleaner if this is only based on the `defaultFilterQueries` in discovery.xml. But that on the other hand has a risk of being forgotten
If there would be a need for multiple configurations containing `UnDiscoverableItems` (e.g. only withdrawn but not filtering on discoverable, or different facets, or type-based), this method would need to be modified again.

Having a separate service method for including withdrawn and private items is actually not specific to this PR, but was already present in the service

mspalti replied:
I agree.  I added this when I saw the `includeUnDiscoverable` boolean available in the public method parameters. 

I wasn't convinced that this step was actually needed and agree that relying on `defaultFilterQueries` in discovery.xml is probably better. 

As someone who has limited experience with solr it's disconcerting to see a query that excludes withdrawn items and then explicitly includes them at a later point. But that magic makes for easier to maintain code at this level 

`q=*:*&fl=dateIssued.year,handle,search.resourcetype,search.resourceid&fq=NOT(withdrawn:true)&fq=NOT(discoverable:false)&fq=search.resourcetype:2&fq=withdrawn:true+OR+discoverable:false&fq=location:me1ba3b9d-d403-4e57-a0d3-a08c782fe427&fq=dateIssued.year:[*+TO+*]&fq=read:(gadf81151-f3b6-4b29-9c09-6f28865fb614+OR+gadf81151-f3b6-4b29-9c09-6f28865fb614)&start=0&rows=1&sort=dateIssued.year_sort+asc`

mspalti replied:
This query was not a valid example. 

As it turns out, the problem is solved by removing the boolean from the search methods. I didn't want to do this at first. But when I discovered that it's used only by the browse system (Jira issue indicates that support for withdrawn/private items needs to be removed from the browse system) I decided this change was in scope.

benbosman commented on dspace-api/src/main/java/org/dspace/discovery/SolrServiceUnDiscoverableItemsPlugin.java:
```diff
@@ -0,0 +1,39 @@
+import org.dspace.core.LogManager;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class SolrServiceUnDiscoverableItemsPlugin implements SolrServiceSearchPlugin {
```
Do we need this additional filter?
Search results will by default verify permissions based on SolrServiceResourceRestrictionPlugin, which should already imply only admins can find withdrawn or non-discoverable items

mspalti replied:
Thanks @benbosman , I'm sure you are right about this.  I'll first verify that the discovery permissions are in place where you indicate and then remove this plugin.  When the Jira issue was written a separate plugin was suggested and this class simply implemented the requirement without verifying that it is needed.

mspalti replied:
Yes, the plugin checks read permissions for all non-system admins.  That presumably makes unnecessary any additional filter to exclude withdrawn and non-discoverable items.  I can also remove filters from the discovery.xml configurations below (that I had tentatively added).

If the pattern holds, this PR will entail only a new discovery.xml configuration and ITs!

mspalti replied:
Although it's not needed for withdrawn items, a plugin appears to be needed for private items.

benbosman replied:
you're right, thinks for this update

benbosman commented on dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoveryConfigurationService.java:
```diff
@@ -69,7 +70,15 @@ public DiscoveryConfiguration getDiscoveryConfiguration(final String name) {
                                                                        final IndexableObject dso) {
-        if (StringUtils.isNotBlank(configurationName) && getMap().containsKey(configurationName)) {
+
+        if (dso instanceof Community && configurationName.contentEquals("showUnDiscoverableItems")) {
```
The normal procedure for defining separate configurations per community or collection is based on the discovery.xml map: https://github.com/DSpace/DSpace/blob/master/dspace/config/spring/api/discovery.xml#L41

mspalti replied:
Yes, I knew this was not the right solution.  I just needed something to address the requirement pending feedback. I see how this is handled in discovery.xml. Will update.

mspalti replied:
The requirement in the Jira issues reads:

> Review the scope/configurationName precedence, a search scoped to a community should look for a configuration named as the configurationName prefix + "." + handle failing back to the configurationName if missing

Having looked at this more, I'm not convinced that this requirement is necessary.  

But, assuming that it is needed, it can only be accomplished by looking for a configuration name in discovery.xml that is a composite of the current configurationName and community handle (from the query scope). That entails something like what was provisionally added here.

Again, it's not clear to me what purpose this serves, and without a scenario to work from it's probably best to ignore this requirement until it's clear why we need it.

benbosman replied:
Agreed

benbosman reviewed (commented):
Thanks @mspalti for the PR
I've included some feedback inline after reading the code.

I wasn't able to test it because my local setup already contains the latest master with changes in spring, flyway, … which caused conflicts.
If you can merge master in your branch, that would be helpful.
There's also an additional discovery config in the latest master which should contain the same defaultFilterQueries

mspalti committed:
Removes browse for private and withdrawn items, removes the includeUnDiscoverable boolean from SearchService methods, adds solr plugin to prevent non-admin discovery of private items.

mspalti commented:
After feedback and more review I made these changes to the PR:

- Removed support for withdrawn and private items from the browse system (as mentioned in the Jira issue)
- Modified the solr `SearchService` by removing the `includeUnDiscoverable` boolean from search methods (this parameter was used only by the browse system)
- Removed my earlier modification of the discovery repository that called the search method using the `includeUnDiscoverable` boolean
- Added a solr plugin to restrict non-admin access to private items (this is not necessary for withdrawn items since these are already protected by an existing plugin).
- Added integration tests.

mspalti commented:
@benbosman , this most recent commit should be up to date with master.

mspalti committed:
Fixed style check error.

tdonohue commented on dspace-api/src/main/java/org/dspace/browse/SolrBrowseDAO.java:
```diff
@@ -227,21 +223,14 @@ private DiscoverResult getSolrResponse() throws BrowseException {
+                query.addFilterQueries(searcher.createLocationQueryForAdministrableItems(context));
             }
+        } catch (SQLException ex) {
+            log.error(ex);
```
While I realize this code isn't new in this PR, can we take this opportunity to improve the log statement here?  I think this should say something like:
`log.error("Error looking up authorization rights of current user", ex);`

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:
```diff
@@ -3582,4 +3582,139 @@ public void discoverSearchObjectsWorkflowConfigurationTest() throws Exception {
+        //An anonymous user browses this endpoint to find the the withdrawn or private objects in the system
+        //With a query stating 'Test'
+        getClient().perform(get("/api/discover/search/objects")
+            .param("configuration", "showUnDiscoverableItems")
```
The name of this configuration was initially confusing to me.  I *thought* that `showUnDiscoverableItems` would return *both* public and undiscoverable items (as the word "show" to me implied it would return both).  However, after looking at the tests here, I realize I was wrong in that assumption...this *only* returns the "undiscoverable" items (excluding the public ones).

That behavior seems OK to me. But, could we rename this configuration slightly to make it clearer?  Maybe call it simply `undiscoverable`, or maybe `restricted`?   That would also align it better with other configurations like `workflow` and `workspace`, which just return Items with that status.

mspalti replied:
Good suggestion.  I was initially thinking in terms of both public and undiscoverable items so the name made sense.  Will change.

I notice that separate endpoints for discovering withdrawn and private items are listed on the estimation spreadsheet, and that there's a question about whether these should be combined into a single endpoint, as is done here. Obviously, if separate configurations are needed that's an easy change. It could be added to this PR if needed.

tdonohue reviewed (commented):
@mspalti : I gave this a quick code review today (haven't tested yet though).  I think it looks good overall, but I have a few minor suggestions inline below.

mspalti committed:
Renamed discovery configuration and added error message.

mspalti committed:
Added check for commuity/collection admins to solr plugin for private items.

mspalti commented:
I modified the `SolrServicePrivateItemPlugin` to allow community and collection admins to discover private items. (BTW, I'm not sure how to write an IT to test for community/collection admins.)

It it's preferable, instead of adding a new plugin for private items, we could add the filter to the exiting `SolrServiceResourceRestrictionPlugin`.

mspalti committed:
Merge remote-tracking branch 'upstream/master' into discoveryAdmin

mspalti committed:
Minor update in discovery.xml

mspalti commented:
Changes from master are now merged.  (Sorry, I thought this was done earlier.)

benbosman commented on dspace-api/src/main/java/org/dspace/discovery/SolrServicePrivateItemPlugin.java:
```diff
@@ -0,0 +1,49 @@
+            // NOTE: the resource restriction plugin adds location filters for community and collection admins.
+            if ( !authorizeService.isAdmin(context) && !authorizeService.isCommunityAdmin(context)
+                && !authorizeService.isCollectionAdmin(context)) {
+                solrQuery.addFilterQuery("NOT(discoverable:false)");
```
Since items which are marked as not discoverable still have public read rights, this implies the SolrServiceResourceRestrictionPlugin would not filter them out (as you discovered in https://github.com/DSpace/DSpace/pull/2580#discussion_r353513281).
This change will imply that collection and community admins will find non-discoverable items with public read rights also outside their collection.
I don't think this is a blocking issue though because:
* This PR already ensures other users won't see the non-discoverable items in discovery
* If the items have read rights, it's probably not that big a deal for comm/coll admins to find them even if they're outside their collection

benbosman reviewed (approved):
I've tested the functionality and noticed this also solves the problem below:

If you make an item private, it was no longer possible to edit the item in the UI using the edit item navigation, not even when starting from the item page, because it's no longer part of the main discovery index. The same occurred if you withdraw an item
With this PR, these problems have been resolved

All the feedback I had included is resolved as well, I only have one minor inline comment which I don't see as a big problem not to resolve (especially since that problem is not created by this PR)

mspalti commented:
@benbosman: Yes, it makes sense that private and withdrawn items are found now - a side benefit of updating the browse system and search service. Thanks for pointing this out!

Perhaps the problem you identified for com/col admins should be looked at, at least enough to determine whether it's a serious concern.  But that seems outside the scope of this PR (unless I've missed something).

tdonohue reviewed (approved):
👍 This looks good to me now. Thanks @mspalti !
